datafusion/physical-expr-common/src/binary_map.rs

```diff
@@ -475,6 +494,106 @@ where
         }
     }
 
+    /// Retrieves the payloads for each value from `values`, either by using
+    /// small value optimizations or larger value handling.
+    ///
+    /// This function will compute hashes for each value and attempt to retrieve
+    /// the corresponding payload from the map. If the value is not found, it will return `None`.
+    ///
+    /// # Arguments:
+    ///
+    /// `values`: The array whose payloads need to be retrieved.
+    ///
+    /// # Returns
+    ///
+    /// A vector of payloads for each value, or `None` if the value is not found.
+    pub fn get_payloads(self, values: &ArrayRef) -> Vec<Option<V>> {
+        match self.output_type {
+            OutputType::Binary => {
+                assert!(matches!(
+                    values.data_type(),
+                    DataType::Binary | DataType::LargeBinary
+                ));
+                self.get_payloads_inner::<GenericBinaryType<O>>(values)
+            }
+            OutputType::Utf8 => {
+                assert!(matches!(
+                    values.data_type(),
+                    DataType::Utf8 | DataType::LargeUtf8
+                ));
+                self.get_payloads_inner::<GenericStringType<O>>(values)
+            }
+            _ => unreachable!("View types should use `ArrowBytesViewMap`"),
+        }
+    }
+
+    /// Generic version of [`Self::get_payloads`] that handles `ByteArrayType`
+    /// (both String and Binary).
+    ///
+    /// See `get_payloads` for more details.
+    fn get_payloads_inner<B>(self, values: &ArrayRef) -> Vec<Option<V>>
+    where
+        B: ByteArrayType,
+    {
+        // Step 1: Compute hashes
+        let mut batch_hashes = vec![0u64; values.len()];
+        batch_hashes.clear();
+        batch_hashes.resize(values.len(), 0);
+        create_hashes(&[values.clone()], &self.random_state, &mut batch_hashes).unwrap(); // Compute the hashes for the values
+
+        // Step 2: Get payloads for each value
+        let values = values.as_bytes::<B>();
+        assert_eq!(values.len(), batch_hashes.len()); // Ensure hash count matches value count
+
+        let mut payloads = Vec::with_capacity(values.len());
+
+        for (value, &hash) in values.iter().zip(batch_hashes.iter()) {
+            // Handle null value
+            let Some(value) = value else {
+                if let Some(&(payload, _)) = self.null.as_ref() {
```

Comment:
I think this code would be faster if you checked for nulls outside the loop rather than on each row